File: tests/Aiko.Core.Tests/FaceArtTests.cs
using System.Text.RegularExpressions;

namespace Aiko.Core.Tests;

public class FaceArtTests
{
    public static TheoryData<FaceStyle, AikoFace, FaceGround, bool> EveryFace()
    {
        var data = new TheoryData<FaceStyle, AikoFace, FaceGround, bool>();
        foreach (var style in Enum.GetValues<FaceStyle>())
        {
            foreach (var face in Enum.GetValues<AikoFace>())
            {
                foreach (var ground in Enum.GetValues<FaceGround>())
                {
                    data.Add(style, face, ground, false);
                    data.Add(style, face, ground, true);
                }
            }
        }

        return data;
    }

    private static readonly HashSet<string> Palette =
        [FaceArt.Ink, FaceArt.Paper, FaceArt.Muted, FaceArt.Caution, FaceArt.Pink, FaceArt.Drop, FaceArt.CautionOnLight];

    // Only the commands the mockup used, with plain numbers: a comma for a decimal point from a
    // Russian Windows would break every face.
    private static readonly Regex PathData = new(@"^[MLHVCQTAZ0-9 .\-]+$");

    [Theory]
    [MemberData(nameof(EveryFace))]
    public void Every_face_is_drawn_in_the_palette_with_readable_path_data(FaceStyle style, AikoFace face, FaceGround ground, bool small)
    {
        var picture = FaceArt.Draw(style, face, ground, small);
        var shapes = picture.Groups.SelectMany(g => g.Shapes).ToList();

        Assert.NotEmpty(shapes);
        Assert.All(shapes, shape =>
        {
            Assert.Matches(PathData, shape.Data);
            Assert.All(shape.Holes ?? [], hole => Assert.Matches(PathData, hole));
            Assert.True(shape.Fill is not null || shape.Stroke is not null);
            Assert.True(shape.Fill is null || Palette.Contains(shape.Fill), shape.Fill);
            Assert.True(shape.Stroke is null || Palette.Contains(shape.Stroke), shape.Stroke);
            Assert.InRange(shape.Opacity, 0.1, 1);
        });
    }

    [Fact]
    public void The_chibi_head_has_an_outline_only_on_a_light_taskbar()
    {
        var dark = FaceArt.Draw(FaceStyle.Chibi, AikoFace.Fresh, FaceGround.Dark, small: true).Groups[0].Shapes[0];
        var light = FaceArt.Draw(FaceStyle.Chibi, AikoFace.Fresh, FaceGround.Light, small: true).Groups[0].Shapes[0];

        Assert.Null(dark.Stroke);
        Assert.Equal(FaceArt.Ink, light.Stroke);
        Assert.Equal(3.26, light.StrokeWidth);
    }

    [Fact]
    public void A_small_chibi_uses_thicker_lines_and_a_tighter_view()
    {
        var big = FaceArt.Draw(FaceStyle.Chibi, AikoFace.Done, FaceGround.Dark, small: false);
        var small = FaceArt.Draw(FaceStyle.Chibi, AikoFace.Done, FaceGround.Dark, small: true);

        Assert.Equal((0, 0, 64), (big.ViewLeft, big.ViewTop, big.ViewSize));
        Assert.Equal((3, 1, 59), (small.ViewLeft, small.ViewTop, small.ViewSize));
        Assert.True(small.Groups[1].Shapes[0].StrokeWidth > big.Groups[1].Shapes[0].StrokeWidth);
    }

    [Fact]
    public void Emoji_highlights_are_holes_so_the_taskbar_shows_through()
    {
        var eyes = FaceArt.Draw(FaceStyle.Emoji, AikoFace.Fresh, FaceGround.Dark, small: true).Groups[0].Shapes
            .Where(s => s.Holes is { Count: > 0 })
            .ToList();

        Assert.Equal(2, eyes.Count);
        Assert.All(eyes, eye => Assert.Equal(FaceArt.Paper, eye.Fill));
    }

    [Theory]
    [InlineData(AikoFace.Tired, FaceArt.Drop)]
    [InlineData(AikoFace.Error, FaceArt.Drop)]
    [InlineData(AikoFace.Waiting, FaceArt.Caution)]
    public void Signs_sit_outside_the_face_group(AikoFace face, string color)
    {
        var picture = FaceArt.Draw(FaceStyle.Chibi, face, FaceGround.Dark, small: true);

        Assert.Contains(picture.Groups[^1].Shapes, s => s.Fill == color);
        Assert.DoesNotContain(picture.Groups[1].Shapes, s => s.Fill == color);
    }

    [Fact]
    public void A_small_emoji_draws_thicker_lines_too()
    {
        var big = FaceArt.Draw(FaceStyle.Emoji, AikoFace.Fresh, FaceGround.Dark, small: false).Groups[0].Shapes.Where(s => s.Stroke is not null).ToList();
        var small = FaceArt.Draw(FaceStyle.Emoji, AikoFace.Fresh, FaceGround.Dark, small: true).Groups[0].Shapes.Where(s => s.Stroke is not null).ToList();

        Assert.Equal(big.Count, small.Count);
        Assert.All(big.Zip(small), pair => Assert.True(pair.Second.StrokeWidth > pair.First.StrokeWidth));
    }
}

File: mac/Tests/AikoKitTests/FaceArtTests.swift
import Foundation
import Testing

@testable import AikoKit

struct FaceArtTests {
    static func everyFace() -> [(FaceStyle, AikoFace, FaceGround, Bool)] {
        var data: [(FaceStyle, AikoFace, FaceGround, Bool)] = []
        for style in FaceStyle.allCases {
            for face in AikoFace.allCases {
                for ground in FaceGround.allCases {
                    data.append((style, face, ground, false))
                    data.append((style, face, ground, true))
                }
            }
        }
        return data
    }

    static let palette: Set<String> = [
        FaceArt.ink, FaceArt.paper, FaceArt.muted, FaceArt.caution, FaceArt.pink, FaceArt.drop, FaceArt.cautionOnLight,
    ]

    // Only the commands the mockup used, with plain numbers: a comma for a decimal point from a
    // Russian system would break every face.
    static let pathData = try! NSRegularExpression(pattern: "^[MLHVCQTAZ0-9 .\\-]+$")

    private func matchesPathData(_ text: String) -> Bool {
        Self.pathData.firstMatch(in: text, range: NSRange(location: 0, length: (text as NSString).length)) != nil
    }

    @Test(arguments: FaceArtTests.everyFace())
    func everyFaceIsDrawnInThePaletteWithReadablePathData(each: (FaceStyle, AikoFace, FaceGround, Bool)) {
        let (style, face, ground, small) = each
        let picture = FaceArt.draw(style, face, ground, small)
        let shapes = picture.groups.flatMap(\.shapes)

        #expect(!shapes.isEmpty)
        for shape in shapes {
            #expect(matchesPathData(shape.data), "path data: \(shape.data)")
            for hole in shape.holes ?? [] {
                #expect(matchesPathData(hole), "hole: \(hole)")
            }
            #expect(shape.fill != nil || shape.stroke != nil)
            #expect(shape.fill == nil || Self.palette.contains(shape.fill!), "\(shape.fill ?? "")")
            #expect(shape.stroke == nil || Self.palette.contains(shape.stroke!), "\(shape.stroke ?? "")")
            #expect(shape.opacity >= 0.1 && shape.opacity <= 1)
        }
    }

    @Test
    func theChibiHeadHasAnOutlineOnlyOnALightBackground() {
        let dark = FaceArt.draw(.chibi, .fresh, .dark, true).groups[0].shapes[0]
        let light = FaceArt.draw(.chibi, .fresh, .light, true).groups[0].shapes[0]

        #expect(dark.stroke == nil)
        #expect(light.stroke == FaceArt.ink)
        #expect(light.strokeWidth == 3.26)
    }

    @Test
    func aSmallChibiUsesThickerLinesAndATighterView() {
        let big = FaceArt.draw(.chibi, .done, .dark, false)
        let small = FaceArt.draw(.chibi, .done, .dark, true)

        #expect(big.viewLeft == 0 && big.viewTop == 0 && big.viewSize == 64)
        #expect(small.viewLeft == 3 && small.viewTop == 1 && small.viewSize == 59)
        #expect(small.groups[1].shapes[0].strokeWidth > big.groups[1].shapes[0].strokeWidth)
    }

    @Test
    func emojiHighlightsAreHolesSoTheBackgroundShowsThrough() {
        let eyes = FaceArt.draw(.emoji, .fresh, .dark, true).groups[0].shapes
            .filter { ($0.holes?.count ?? 0) > 0 }

        #expect(eyes.count == 2)
        for eye in eyes {
            #expect(eye.fill == FaceArt.paper)
        }
    }

    @Test(arguments: [
        (AikoFace.tired, FaceArt.drop),
        (AikoFace.error, FaceArt.drop),
        (AikoFace.waiting, FaceArt.caution),
    ])
    func signsSitOutsideTheFaceGroup(face: AikoFace, color: String) {
        let picture = FaceArt.draw(.chibi, face, .dark, true)

        #expect(picture.groups[picture.groups.count - 1].shapes.contains { $0.fill == color })
        #expect(!picture.groups[1].shapes.contains { $0.fill == color })
    }

    @Test
    func aSmallEmojiDrawsThickerLinesToo() {
        let big = FaceArt.draw(.emoji, .fresh, .dark, false).groups[0].shapes.filter { $0.stroke != nil }
        let small = FaceArt.draw(.emoji, .fresh, .dark, true).groups[0].shapes.filter { $0.stroke != nil }

        #expect(big.count == small.count)
        for (first, second) in zip(big, small) {
            #expect(second.strokeWidth > first.strokeWidth)
        }
    }
}
